File: win_avoidability.py
from chessball_board import ChessBallBoard, Player, possible_previous_moves
from blocking_move import find_blocking_move

def is_win_avoidable_by_opponent(position: ChessBallBoard, player: Player) -> bool:
    """
    Given a position in which `player` has a winning move,
    checks whether in ALL possible previous positions (and moves for the opponent)
    there exists a blocking move for the opponent that prevents the win.
    In other words: returns True if the opponent could have always blocked this win,
    and False if some previous position made the win inevitable.
    """
    opponent = Player.BLACK if player == Player.WHITE else Player.WHITE
    previous_positions = possible_previous_moves(position, opponent)
    if not previous_positions:  # No prior position exists
        return False
    for move, prev_board in previous_positions:
        blocking = find_blocking_move(prev_board, opponent)
        if blocking is None:
            # At least one prior position did NOT allow a blocking move: win is inevitable
            return False
    # All previous positions allowed a blocking move for the opponent
    return True

File: rust_chessball/src/win_avoidability.rs
//! Check whether a win was avoidable by the opponent by examining possible previous positions.

use crate::blocking_move::find_blocking_move;
use crate::board::{ChessBallBoard, Player};
use crate::moves::possible_previous_moves;

/// Given a `position` in which `player` has a winning move, checks whether in ALL possible previous
/// positions there existed a blocking move for the opponent that prevented the win.
///
/// Returns `true` if the opponent could have always blocked (i.e., the win was avoidable),
/// and `false` if there exists a previous position that made the win inevitable.
pub fn is_win_avoidable_by_opponent(position: &ChessBallBoard, player: Player) -> bool {
    let opponent = match player {
        Player::White => Player::Black,
        Player::Black => Player::White,
        Player::Neutral => Player::Neutral,
    };
    let previous_positions = possible_previous_moves(position, opponent);
    if previous_positions.is_empty() {
        return false;
    }
    for (_mv, prev_board) in previous_positions {
        let blocking = find_blocking_move(&prev_board, opponent);
        if blocking.is_none() {
            return false;
        }
    }
    true
}
